master/custom/release_dashboard.py

```diff
@@ -1,122 +1,730 @@
+            return '\N{CIRCLED TIMES}'
+        return str(self["results"])
 
-        if cache is not None and not force_refresh:
-            result, deadline = cache
-            if time.monotonic() <= deadline:
-                return result
+    @cached_property
+    def results_string(self):
+        return buildbot.process.results.statusToString(self["results"])
 
-        result = get_release_status()
-        deadline = time.monotonic() + CACHE_DURATION
-        cache = (result, deadline)
+    @cached_property
+    def css_color_class(self):
+        if self["results"] == buildbot.process.results.SUCCESS:
+            return 'success'
+        if self["results"] == buildbot.process.results.WARNINGS:
+            return 'warning'
+        if self["results"] == buildbot.process.results.FAILURE:
+            return 'danger'
+        return 'unknown'
+
+    @cached_property
+    def junit_results(self):
+        filepath = (
+            self._root._app.test_result_dir
+            / self.builder.branch.tag
+            / self.builder["name"]
+            / f'build_{self["number"]}.xml'
+        )
+        try:
+            file = filepath.open()
+        except OSError:
+            return None
+        with file:
+            etree = ElementTree.parse(file)
+        result = JunitResult(self, {})
+        for element in etree.iterfind('.//error/..'):
+            result.add(element)
         return result
 
-    return release_status_app
+    @cached_property
+    def duration(self):
+        try:
+            seconds = (
+                self["complete_at"]
+                - self["started_at"]
+                - self["locks_duration_s"]
+            )
+        except (KeyError, TypeError):
+            return None
+        return datetime.timedelta(seconds=seconds)
+
+
+class JunitResult(DashboardObject):
+    def __init__(self, *args):
+        super().__init__(*args)
+        self.contents = {}
+        self.errors = []
+        self.error_types = set()
+
+    def add(self, element):
+        errors = []
+        error_types = set()
+        for error_elem in element.iterfind('error'):
+            new_error = JunitError(self, {
+                **error_elem.attrib,
+                'text': error_elem.text,
+            })
+            errors.append(new_error)
+            error_types.add(new_error["type"])
+        result = self
+        name_parts = element.attrib.get('name', '??').split('.')
+        if name_parts[0] == 'test':
+            name_parts.pop(0)
+        for part in name_parts:
+            result.error_types.update(error_types)
+            result = result.contents.setdefault(part, JunitResult(self, {}))
+        result.error_types.update(error_types)
+        for error in errors:
+            if error not in result.errors:
+                # De-duplicate, since failing tests are re-run and often fail
+                # the same way
+                result.errors.extend(errors)
```

Comment:
Hummmm aren'y you adding technically all the errors again and again once per error to the list? What am I missing?